### PR TITLE
Clean up sound_fade

### PR DESCRIPTION
**Pull request by hecktest**
`minetest.sound_fade` has horrible documentation and even worse code:
- No mention of the units that `step` is in
- No mention of zero gain quietly deleting a sound (!)
- Asking for a _signed speed_ while also specifying a target gain (aaaaaaaa _why_)
- Easy to bug out by specifying a mismatching speed and target, or zero speed (duh)
- Pegged to 10 frames per second, resulting in audible artifacts. Undocumented, of course.

And even when I followed the stupid parameter conventions, I couldn't get it to work properly. So I'm not entirely sure if this code _ever_ worked, doesn't seem like many people are using it.

This PR does:
- Make the step parameter unsigned (compatible with old code via abs)
- Document the function properly
- Handle edge cases (suggestions welcome here, can be anything as long as we at least _define_ the behavior)
- Remove the cringeworthy 10 fps "optimization" (you're welcome to argue about this... and lose)
- Move the gain to target exactly. (I think the original overshoots it)
- Clamp when receiving the gain, not every time it's moved.
- Make the intent of the C++ code clear (I'll defend this to death, the original is probably the most convoluted "move towards" function I've seen in my life.)

### CODE REVIEW

**Comment by paramat** [comment]
Hm the previous implementatin does seem bad.
Fading sound in 0.1s steps will be audibly stepped instead of a smooth fade, a second is a long time.
Having 'step' and 'target gain' is nasty, it should obviously have been 'target gain' and 'fade time', from which fade speed is calculated.

**Comment by paramat** [comment]
Heh:
https://github.com/minetest/minetest/pull/5340#issuecomment-298493486
"The fade is choppy, with clear volume steps. I do not recommend fading sounds quickly"
But was still approved.

**Comment by appgurueu** [comment]
A quick Google search confirms that there are likely no usages: https://www.google.com/search?q=%22minetest.sound_fade%22

**Comment by hecktest** [comment]
Bump?
Please review.

**Comment by SmallJoker** [comment]
Test script N°83, best take some ambient music.
```Lua
local handle
minetest.register_on_joinplayer(function(player)
	minetest.after(1, function()
		handle = minetest.sound_play({
			name = "sax_violin", -- custom, long sound here
			gain = 1.0
		}, {
			to_player = player,
			gain = 0.2,
			loop = true
		})
	end)
end)

minetest.register_chatcommand("s", {
	description = "Modify sound. <step> <gain>",
	privs = {},
	func = function(name, param)
		param = param:split(" ")
		local step = tonumber(param[1]) or 0
		local gain = tonumber(param[2]) or 0
		minetest.sound_fade(handle, step, gain);
	end
})
```

**Comment by SmallJoker** [comment]
Interestingly the sound can no longer be faded out when `step` was set to zero once. Perhaps a bug to fix in the future?

File: src/client/sound_openal.cpp [hunk 1]
```diff
@@ -616,38 +614,40 @@ class OpenALSoundManager: public ISoundManager
...
 	}
 
 	void doFades(float dtime)
 	{
-		m_fade_delay += dtime;
-
-		if (m_fade_delay < 0.1f)
-			return;
+		for (auto i = m_sounds_fading.begin(); i != m_sounds_fading.end();) {
+			i->second.current_gain += (i->second.step * dtime);
```
Comment (sfan5) [inline src/client/sound_openal.cpp:633]: A `FadeState &s = i->second;` before here would improve the readability of the rest

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5186,9 +5186,10 @@ Sounds
 * `minetest.sound_fade(handle, step, gain)`
     * `handle` is a handle returned by `minetest.sound_play`
     * `step` determines how fast a sound will fade.
-      Negative step will lower the sound volume, positive step will increase
-      the sound volume.
+      The gain will change by this much per second,
+      until it reaches the target gain.
```
Comment (sfan5) [inline doc/lua_api.txt:5190]: should have a note along the lines of
"Note that prior to Minetest 5.4.0 the sign of `step` had to be negative to fade to a lower volume"

**Comment by hecktest** [comment]
> Interestingly the sound can no longer be faded out when `step` was set to zero once. Perhaps a bug to fix in the future?
What do you mean? It's not possible to set the step to zero now, because the client will just discard that command. Even if one were to force a step of zero somehow, it would not prevent the fade from being replaced by a new, valid command. (Fades are not queued, they're replaced.) Did I miss something here?

**Comment by hecktest** [comment]
All concerns hopefully addressed.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**